formats/json-tests/jvmTest/src/kotlinx/serialization/JavaCollectionsTest.kt

```diff
@@ -49,4 +51,55 @@ class JavaCollectionsTest {
         val restored = Json.decodeFromString(deserializer = serializer, string = string)
         assertEquals(expected = original, actual = restored)
     }
+
+    @Test
+    fun testTopLevelMaps() {
+        // Returning null here is a deliberate choice: map constructor functions may return different specialized
+        // implementations (e.g., kotlin.collections.EmptyMap or java.util.Collections.SingletonMap)
+        // that may or may not be generic. Since we generally cannot return a generic serializer using Java class only,
+        // all attempts to get map serializer using only .javaClass should return null.
+        assertNull(serializerOrNull(emptyMap<String, String>().javaClass))
+        assertNull(serializerOrNull(mapOf<String, String>("a" to "b").javaClass))
+        assertNull(serializerOrNull(mapOf<String, String>("a" to "b", "b" to "c").javaClass))
+        // Correct ways of retrieving map serializer:
+        assertContains(
+            serializer(typeTokenOf<Map<String, String>>()).descriptor.serialName,
+            "kotlin.collections.LinkedHashMap"
+        )
+        assertContains(
+            serializer(typeTokenOf<java.util.LinkedHashMap<String, String>>()).descriptor.serialName,
+            "kotlin.collections.LinkedHashMap"
+        )
+        assertContains(
+            serializer(typeOf<LinkedHashMap<String, String>>()).descriptor.serialName,
+            "kotlin.collections.LinkedHashMap"
+        )
+    }
+
+    @Test
+    fun testTopLevelSetsAndLists() {
+        // Same reasoning as for maps
+        assertNull(serializerOrNull(emptyList<String>().javaClass))
+        assertNull(serializerOrNull(listOf<String>("a").javaClass))
+        assertNull(serializerOrNull(listOf<String>("a", "b").javaClass))
+        assertNull(serializerOrNull(emptySet<String>().javaClass))
+        assertNull(serializerOrNull(setOf<String>("a").javaClass))
+        assertNull(serializerOrNull(setOf<String>("a", "b").javaClass))
+        assertContains(
+            serializer(typeTokenOf<Set<String>>()).descriptor.serialName,
```

Comment:
Should we also add a test on KClass for the `serializer(kClass: KClass<*>, typeArgumentsSerializers: List<KSerializer<*>>, isNullable: Boolean)` function?

Reply:
It's a debatable question. Currently, `serializer(List::class, listOf(String.serializer()), false)` returns the proper serializer for a list of strings. However, `serializer(emptyList<String>()::class, listOf(String.serializer()), false)` does not because we check strict equality for `List::class`, not subtyping. From an ideological point of view, I'd say that it could return a proper serializer, so returning `null` for now is a bug/implementation detail, not a dedicated design choice. So it's different from `serializerOrNull(emptyMap<String, String>().javaClass)`, where it is specified by design that we return `null` on generic classes (even if their implementation is non-generic).